docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+For example, I think that in general we should avoid returning an abstract trait to the user (intermediate traits like Iterator might be fine though since a user is expected to consume those fairly immediately).
+
+# Idiomatic Code
+
+A lot of this section just boils down to "[Read the Book](https://doc.rust-lang.org/book/)", but I'm highlighting a few things I personally found useful while learning.
+
+## Pass by Reference vs Pass by Value
+
+There is something a bit subtle about how Rust handles "pass by value" which means you should not just apply standard C++ best-practice. In particular, in C++ you would might expect a method like:
+
+```
+ReturnType ClassName::FunctionName(ParamType param) { … }
+```
+
+to copy the parameter and return types during the course of calling the method (e.g. as a memcpy to the stack). This leads C++ developers to favour passing objects by reference and passing pointers to hold return types, instead of taking the hit of copying the value.
+
+Rust treats "pass by value" a little differently, and it can be thought more of an indication that "ownership is being transferred" rather than the API wants "a copy of the data". Thus the Rust equivalent code:
+
+```
+fn function_name(param: ParamType) -> ReturnType { … }
+```
+
+may still have parameters "passed by reference" when it can determine that the reference is no longer subsequently used by the calling code (this is a "move" in Rust parlance).
+
+Furthermore, it will often [ellide the copy[(https://en.wikipedia.org/wiki/Copy_elision) of the return value if it determines the returned object would otherwise go out of scope. It will allocate the space for the return value on the caller's stack or use the memory in a destination struct, to directly write the "returned" value in its final destination with no copying whatsoever. This is called Return Value Optimization (RVO) and while it is now available in C++ as well, it's a relatively new feature there.
+
+It is still often better (for reasons of borrowing and ownership) to pass structs by non-mutable reference, but returning **newly created** results by value (even potentially large structures) is not expected to cause performance issues.
+
+### Return new instance by value :: suggested
+
+Use "return by value" for any functions which produce new data and let Rust handle optimizing away any copying, unless there's really measurable performance issue.
+
+### Pass struct parameters by reference where possible :: suggested
+
+When passing struct types to functions, it's always semantically safe to pass a reference (since there can be no race conditions around its use and it cannot be modified unexpectedly while being processed). The called function should be responsible for taking a copy if it needs to (e.g. by cloning or via the ToOwned trait).
+
+### Prefer passing Option<T> by value :: suggested
```

Comment:
this doesn't work if `T` isn't `Copy`, sometimes you need to pass `Option<&T>` or `&Option<T>`. I usually go with the latter since you can get the former with `.as_ref()`.

Reply:
Hmm, so there's some discussion around "Copy + Option" vs "non-Copy + &Option<T>".
I'd like to understand a bit more about the trade-offs here, but that's mostly for my personal understanding. For now I'm happy to just accept any preferred wording changes here.